### PR TITLE
Integration tests

### LINKED ISSUE #124: Consensus testing of minscripts
#123 brought an interesting point that we should consensus test our miniscript with rust-libbitcoinconsensus. 

Create transactions using rust-miniscripts and check the transaction validity/satisfaction logic/weight calculation etc and other features with https://github.com/rust-bitcoin/rust-bitcoinconsensus. We can also grammar that Pieter had to generate random miniscirpts. 

Not completely sure of how we are going to do this, but I think we should have this.

### PR DESCRIPTION
**Pull request by sanket1729**
Added integration tests based for consensus and standardness checks for bitcoin core 0.20.0. Downloads the `bitcoind` binary form bitcoin.org and uses `rust-bitcoin-core-rpc` crate to make rpc queries to it. This serves as a test of our current satisfier logic as well as the rust-bitcoin psbt against the bitcoin core consensus and the policy rules. 

Note that we could not use the rust wrapper of `libbitcoinconsensus` because it does not expose the standardness flags. The current integration test setup should also help us test upcoming consensus changes in Taproot upgrade. Currently, we only test consensus/policy rules against bitcoin core 0.20.0 but I think that should be sufficient. 

Fixes #124 .

Marked as draft because some cases are silently failing because we cannot yet satisfy psbt hash preimages.

### CODE REVIEW

**Comment by sanket1729** [comment]
The CI fails should be resolved once #161 is merged.

File: contrib/test.sh
Comment (RCasatta) [inline contrib/test.sh:75]: with curl you can pipe directly to avoid saving an external file (and removing it), something like:
https://bitcoincore.org/bin/bitcoin-core-$BITCOINVERSION/bitcoin-$BITCOINVERSION-x86_64-linux-gnu.tar.gz | tar -xvz 

optionally to speed up a bit you can also extract only the bitcoind executable, but in that case you have to change the shasum verification part so maybe not worthy

Comment (sanket1729) [reply contrib/test.sh]: Thanks, updated the PR with both suggestions.

File: integration_test/src/test_random_miniscripts.rs
Comment (apoelstra) [inline integration_test/src/test_random_miniscripts.rs:24]: i think we should read from stdin

File: integration_test/src/test_random_miniscripts.rs
Comment (apoelstra) [inline integration_test/src/test_random_miniscripts.rs:21]: should be an iterator rather than reading everything at once into a vec

File: integration_test/Cargo.toml
Comment (apoelstra) [inline integration_test/Cargo.toml:8]: I wonder .... if we can use the p2p layer rather than RPC here

File: contrib/test.sh
Comment (shesek) [reply contrib/test.sh]: This could also be done with `wget -O - ... | tar ...`

**Comment by sanket1729** [comment]
Resuming work on this. Now that we have rust-bitcoin 0.26

File: integration_test/src/test_random_miniscripts.rs
Comment (sanket1729) [reply integration_test/src/test_random_miniscripts.rs:20]: So, the tradeoff of avoiding allocations is that we will need to mine a single transaction per block to see that it is getting confirmed. If we collect all the miniscripts into a `Vec`. We can do the following. 
1) Create transactions spending to these Miniscripts.
2) Wait for confirmation
3) Spend from the Miniscripts
4) More confirmations

I guess I can mine 1 tx per block. Not a big deal.

**Comment by sanket1729** [comment]
Based on #227. Still awaiting a release of `rust-bitcoincore-rpc` with 0.26 types.

**Comment by sanket1729** [comment]
This is also ready for review. The CI would show incomplete because this PR changes the CI script

File: contrib/test.sh
Comment (apoelstra) [inline contrib/test.sh:56]: nit: trailing whitespace here

File: contrib/test.sh
Comment (apoelstra) [inline contrib/test.sh:58]: I think you're forgetting to download the sha256sum? At least,when I try running this with 0.21.1 or 0.20.1 it doesn't work here.

Comment (sanket1729) [reply contrib/test.sh:58]: The file would have to be manually downloaded, sha256sum checked locally, and then finally committed here. Unfornutealy, this will be a manual process for every release of bitcoin core. 

I don't see the point of downloading the sha2 sum from bitcoincore.org if we are downloading the binary from there :P .

Comment (apoelstra) [reply contrib/test.sh:58]: Ah, perfect, I understand. I don't know how I didn't see the sha256sum file checked in for 0.20.0 (now 0.21.0).

File: contrib/test.sh
Comment (apoelstra) [reply contrib/test.sh:56]: This nit is still here

File: integration_test/src/main.rs
Comment (apoelstra) [inline integration_test/src/main.rs:52]: As near as I can tell there is no reason for this function or the `static mut` global to exist

File: contrib/test.sh
Comment (sanket1729) [reply contrib/test.sh:58]: Will update the version to 0.21.0

**Review by sanket1729 — COMMENTED** [review]
This would be blocked on https://github.com/rust-bitcoin/rust-bitcoincore-rpc/pull/196 for the time being

File: integration_test/Cargo.toml
Comment (RCasatta) [inline integration_test/Cargo.toml:9]: outdated comment

**Review by RCasatta — COMMENTED** [review]
FYI, maybe for future PR, I think [`bitcoind`](https://crates.io/crates/bitcoind) crate as a dev dep could help to:

* avoid shell scripts
* potentially use different bitcoind version by just changing a feature
* make multiple independent, parallel tests (in bdk for example every #[test] spawn its own node)
* have guarantees bitcoind process are terminated whether the test terminate correctly or fail, while avoiding the killall that on local machine could terminate other bitcoind process on the dev machine

I think there is an issue with bitcoind MSRV which is now 1.41.1, but I may look to lower that if there is interest to use it

Disclaimer: I did the crate so maybe I am biased :)

**Comment by sanket1729** [comment]
@RCasatta I read through the features and I really like it. I definitely want to migrate to this because of all the reasons you listed. 

> I think there is an issue with bitcoind MSRV which is now 1.41.1, but I may look to lower that if there is interest to use it

This is fine because at least for our CI, we are planning to run integration tests on rust stable. I will be on my TODO list after we have taproot support in the rust bitcoin ecosystem.

File: integration_test/run.sh
Comment (apoelstra) [inline integration_test/run.sh:3]: In dd322eb8fb9d7732b4e6ec7ca8c0213763ba4200:

This is fine for now but we should use a real single-use temp name. This is going to conflict e.g. whenever you and I are simultaneously running these tests on my big server.

**Review by apoelstra — APPROVED** [review]
ACK 712ffbe907529605fcc9d0757012e4bdf1b4017d

We can leave the nits to a followup PR.

**Comment by sanket1729** [comment]
Merging this. Created an issue to address Andrew's comment. #287

**Pull request merged**
